mountpoint-s3-crt/src/s3/client.rs

```diff
@@ -570,54 +571,90 @@ unsafe impl Send for MetaRequest {}
+        Self {
+            request,
+            slice,
+            eof,
+            waker: Default::default(),
         }
     }
 }
 
-impl<'a> MetaRequestWrite<'a> {
-    fn new(request: &'a mut MetaRequest, slice: &'a [u8], eof: bool) -> Self {
-        // SAFETY: `MetaRequestWrite` will ensure that `slice` is alive until the future completes or
-        // that the meta-request is canceled if the future is dropped, preventing further use of the
-        // `aws_byte_cursor`.
-        let data = unsafe { slice.as_aws_byte_cursor() };
-
-        // SAFETY: `aws_s3_meta_request_write` never returns NULL.
-        let future = unsafe {
-            FutureVoid::from_crt(NonNull::new_unchecked(aws_s3_meta_request_write(
-                request.inner.as_ptr(),
+impl<'r, 's> Future for MetaRequestWrite<'r, 's> {
+    type Output = Result<&'s [u8], Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Self::Output> {
+        let mut waker = self.waker.lock().unwrap();
+        if let Some(ref mut waker) = *waker {
+            // The write is still pending. Make sure we store the waker from the current context.
+            waker.clone_from(cx.waker());
+            return std::task::Poll::Pending;
+        }
+
+        // Store the waker.
+        *waker = Some(cx.waker().clone());
+
+        // `user_data` will be dropped in `poll_write_waker_callback` (or below).
+        let user_data = Arc::into_raw(self.waker.clone()) as *mut ::libc::c_void;
+
+        // SAFETY: `aws_s3_meta_request_poll_write` does not store `data`.
+        let data = unsafe { self.slice.as_aws_byte_cursor() };
+
+        // SAFETY: `self.request` wraps a valid `aws_s3_meta_request` pointer.
+        let result = unsafe {
+            aws_s3_meta_request_poll_write(
+                self.request.inner.as_ptr(),
                 data,
-                eof,
-            )))
+                self.eof,
+                Some(poll_write_waker_callback),
+                user_data,
+            )
         };
+        if result.is_pending {
+            return std::task::Poll::Pending;
+        }
 
-        Self { future, request }
-    }
-}
+        // SAFETY: `aws_s3_meta_request_poll_write` completed. It will not invoke `poll_write_waker_callback`,
+        //         so we need to drop `user_data` here.
+        _ = unsafe { Arc::from_raw(user_data as *mut Mutex<Option<Waker>>) };
 
-impl Future for MetaRequestWrite<'_> {
-    type Output = Result<(), Error>;
+        let error_result: crate::common::error::Error = result.error_code.into();
+        let result = if error_result.is_err() {
+            Err(error_result)
+        } else {
+            Ok(&self.slice[result.bytes_processed..])
+        };
 
-    fn poll(mut self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Self::Output> {
-        Pin::new(&mut self.future).poll(cx)
+        std::task::Poll::Ready(result)
```

Comment:
may it happen that an executor will continue polling a future after it's ready and we end up writing the same slice twice?

i've noticed that we use `.fuse()` in [other code paths](https://github.com/awslabs/mountpoint-s3/blob/main/mountpoint-s3/src/inode.rs#L724) _seemingly_ to avoid that: 
```
let mut file_lookup = client.head_object(&self.bucket, &full_path).fuse();
```

from fuse method doc:
```
    /// Fuse a future such that `poll` will never again be called once it has
    /// completed. This method can be used to turn any `Future` into a
    /// `FusedFuture`.
```

it does not seem that we have this guardrail on the [write path](https://github.com/awslabs/mountpoint-s3/blob/main/mountpoint-s3/src/upload.rs#L166):
```
self.request.write(data).await?;
```

Reply:
Correct, this is not a `FusedFuture`. That is required in some use cases (e.g. with `select!`), but that's what `.fuse()` is for. I don't think it is a concern here.